### DISCUSSION TITLE
Why is recall rate decreasing for increasing input vector (nq) for IVF algorithms only?

### DISCUSSION THREAD

**Discussion by nilsertle** [milvus-io/milvus#25503] [Q&A and General discussion]
I have tested the Milvus database for some configurations and ploted the qps (queries per second) over the recall@1 rate for different nqs (nq=1, nq=10, nq=100). The point marked with a black circle for each line is the nq=1 test.

Why is the recall rate decreasing for all IVF algorithms but not really for the other? And why is it decreasing at all? I used the AnnSift1M dataset. 

![qps_vs_recall@1_for_all_nq](https://github.com/milvus-io/milvus/assets/89461027/e844214f-bf9d-4b52-8015-f3c2fc68bf96)

**Comment by yhmo** [comment]
For IVF index, nq=100, nprobe=8, the search engine does the following things:
- there is a thread pool, each thread handles one vector
- thread_1 finds out 8 nearest buckets for this vector_1, and calculates the distance to find out topk most similar vectors
- thread_2 finds out 8 nearest buckets for this vector_2, and calculates the distance to find out topk most similar vectors
- ......
- thread_n finds out 8 nearest buckets for this vector_n, and calculates the distance to find out topk most similar vectors

So, theoretically, The recall rate of nq=100 is equal to searching 100 times with nq=1.
To compare the recall rate, you should make sure the 100 vectors are the same, no matter query with nq=100 or nq=1 for 100 times.

**Reply by nilsertle** [reply]
Indeed, they are the same vectors. Therefore, you are suggesting that the results should have the same recall rate for one algorithm regardless of whether it has a nq=1 or nq=100?

What could have been a possible error?

Where is it stated that IVF indexes work in that way? Is there a theory I can read more about? This sounds very interesting and I would like to include it in my thesis.

**Reply by yhmo** [reply]
A possible reason is: there are some buckets that have the same distance to the query vector. With a small nprobe value, let's say 8, the search engine might find out different 8 buckets in parallel computing.


The IVF index algorithm is simple. Milvus search engine implements a multi-thread model to improve the performance by parallel computing. We don't have articles for the implementation, the search engine repo is here: https://github.com/milvus-io/knowhere
